src/eigsolve/arnoldi.jl

```diff
@@ -268,33 +286,30 @@ The return value is always of the form `vals, vecs, info = eigsolve(...)` with
     if `info.converged >= howmany`.
 """
 function realeigsolve(A, x₀, howmany::Int, which::Selector, alg::Arnoldi; alg_rrule=alg)
-    T, U, fact, converged, numiter, numops = _schursolve(A, x₀, howmany, which, alg)
-    if !(eltype(T) <: Real)
-        throw(ArgumentError("realeigsolve can only be used for real eigenvalue problems"))
-    else
-        allreal = true
-        for i in 1:(howmany < length(fact) ? howmany : howmany - 1)
-            if T[i + 1, i] != 0
-                allreal = false
-                break
-            end
-        end
-        allreal || throw(ArgumentError("not all first `howmany` eigenvalues are real"))
-    end
-    if converged > howmany
-        while howmany < converged && T[howmany + 1, howmany] == 0
-            howmany += 1
+    realinner = real ∘ inner
+    v₀ = InnerProductVec(x₀, realinner)
+    realA = v -> InnerProductVec(apply(A, v[]), realinner)
```

Comment:
How would you feel about avoiding the anonymous functions:
```julia
const realinner = real ∘ inner
const RealVec{V} = InnerProductVec{typeof(realinner),V}
RealVec(v) = InnerProductVec(v, realinner)

apply(A, x::RealVec) = RealVec(apply(A, x[]))
...
```
The main point being that the KrylovKit methods then automatically use the correct functions

(I don't care too much about the names, mostly about the interface)

Reply:
I am definitely in favour of the first three lines. The redefinition of `apply` might be a bit more problematic, as maybe some user is constructing the same type of `RealVec` objects, as this definition of `real ∘ inner` does not create a new function that is intrinsic and private to KrylovKit.jl. That user might then define their map to unpack the vector, and this would lead to a conflict. 

The solution might be to have an internal `_realinner(v,w) = real(inner(v,w))` such that `RealVec{V} = InnerProductVec{typeof(_realinner, V}` is impossible to ever be constructed by an external user, without hijacking our private `_realinner`.

Reply:
that's definitely also fine by me